book/src/rfcs/foreign_field_mul.md

```diff
@@ -0,0 +1,1273 @@
+
+All that remains is to range check $v_0$ and $v_1$
+> 7. Range check $v_0 \in [0, 2^2)$
+> 8. Range check $v_1 =\in [0, 2^{\ell + 3})$
+
+## Subtractions
+
+When unconstrained, computing $u_0 = p_0 + 2^{\ell} \cdot p_{10} - (r_0 + 2^{\ell} \cdot r_1)$ could require borrowing.  Fortunately, we have the constraint that the $2\ell$ least significant bits of $u_0$ are `0` (i.e. $u_0 = 2^{2\ell} \cdot v_0$), which means borrowing cannot happen.
+
+Borrowing is prevented because when the the $2\ell$ least significant bits of the result are `0` it is not possible for the minuend to be less than the subtrahend.  We can prove this by contradiction.
+
+Let
+
+* $x = p_0 + 2^{\ell} \cdot p_{10}$
+* $y = r_0 + 2^{\ell} \cdot r_1$
+* the $2\ell$ least significant bits of $x - y$ be `0`
+
+Suppose that borrowing occurs, that is, that $x < y$.  Recall that the length of $x$ is $2\ell + 2$ bits.  Therefore, since $x < y$ the top two bits of $x$ must be zero and so we have
+
+$$
+x - y = x_{2\ell} - y,
+$$
+
+where $x_{2\ell}$ denotes the $2\ell$ least significant bits of $x$.
+
+Recall also that the length of $y$ is $2\ell$ bits.  So the result of this subtraction is $2\ell$ bits.   Since the $2\ell$ least significant bits of the subtraction are `0` this means that
+
+$$
+\begin{aligned}
+x - y  &  = 0 \\
+x &= y,
+\end{aligned}
+$$
+
+which is a contradiction with $x < y$.
+
+## Costs
+
+Range checks should be the dominant cost, let's see how many we have.
+
+Range check (3) requires two range checks for $p_{11} = p_{111} \cdot 2^\ell + p_{110}$
+ * a) $p_{110} \in [0, 2^\ell)$
+ * b) $p_{111} \in [0, 2^2)$
+
+Range check (8) requires two range checks and a decomposition check that is merged in (6).
+ * a) $v_{10} \in [0, 2^{\ell})$
+ * b) $v_{11} \in [0, 2^3)$
+
+The range checks on $p_0, p_1$ and $p_2$ follow from the range checks on $a,b$ and $q$.
+
+So we have 3.a, 3.b, 4, 7, 8.a, 8.b.
+
+| Range check | Gate type(s)                               | Witness                   | Rows |
+| ----------- | ------------------------------------------ | ------------------------- | ---- |
+| 7           | $(v_0 - 3)(v_0 - 2)(v_0 - 1)v_0$           | $v_0$                     | < 1  |
+| 3.a         | $(p_{111} - 3)(p_{111} - 2)(p_{111} - 1)q$ | $p_{111}$                 | < 1  |
+| 8.b         | degree-8 constraint or plookup             | $v_{11}$                  | 1    |
+| 3.b, 4, 8.a | `multi-range-check`                        | $p_{10}, p_{110}, v_{10}$ | 4    |
+
+So we have 1 multi-range-check, 1 single-range-check and 2 low-degree range checks. This consumes just over 5 rows.
+
+## Use CRT to constrain $a \cdot b - q \cdot f - r \equiv 0 \mod n$
+
+Until now we have constrained the equation $\mod 2^t$, but remember that our application of CRT means that we must also constrain the equation $\mod n$.  We are leveraging the fact that if the identity holds for both moduli in $M = \{n, 2^t\}$, then it holds for $\mathtt{lcm} (M)$.
```

Comment:
I think you want to say lcm(n, 2^t) = M, but in any case this is only true because gcd(n, 2^t) = 1, otherwise M wouldn't be the product of n and 2^t.

Reply:
This makes me wonder if ffmul as we design it would also work if n was a power of two and not just a prime.

Reply:
Good spot - my use of M there was inconsistent with previous usage

Reply:
> This makes me wonder if ffmul as we design it would also work if n was a power of two and not just a prime.
Because then $\gcd(n, 2^t) \ne 1$, right?

Reply:
That's why, since they are not coprime anymore then I don't know if the claim about the product would still hold, perhaps it becomes a claim about their lcm, which will not be 2^t * n but 2^t (the largest between them).

Reply:
Oh, I think the moduli still need to be coprime. Maybe I should have restated that here.